Change `FrameTree::uniqueName()` to generate a unique name when called

charliewolfe opened the pull request:
#### b903f425a02b341d81be1131835fd495ae5e100c
<pre>
Change `FrameTree::uniqueName()` to generate a unique name when called
<a href="https://bugs.webkit.org/show_bug.cgi?id=272062">https://bugs.webkit.org/show_bug.cgi?id=272062</a>
<a href="https://rdar.apple.com/125808920">rdar://125808920</a>

Reviewed by Alex Christensen.

Unique frame names currently use a counter to increment a unique frame index each time
`FrameTree::setSpecifiedName` is called. This will not work with site isolation. Instead, we should
change `FrameTree::uniqueName()` to create the unique frame name based on the frame’s position in the
frame tree to mimic previous behavior . Other than layout test output, there are two users of unique
frame names. `HistoryItem` relies on them for session state and to give a consistent identifier for
frame positions after a navigation. Web archives output them in their saved state. Both should continue
to work after this change.

Some tests that output unique frame names needed to be rebased to accommodate this change.

* LayoutTests/fast/forms/form-and-frame-interaction-retains-values-expected.txt:
* LayoutTests/fast/frames/iframe-reparenting-unique-name-expected.txt:
* LayoutTests/fast/frames/reparent-in-unload-contentdocument-expected.txt:
* LayoutTests/http/tests/security/dataURL/xss-DENIED-from-data-url-in-foreign-domain-subframe-expected.txt:
* LayoutTests/http/tests/site-isolation/notify-done-expected.txt:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/page/FrameTree.cpp:
(WebCore::FrameTree::setSpecifiedName):
(WebCore::FrameTree::clearName):
(WebCore::FrameTree::childBySpecifiedName const):
(WebCore::FrameTree::find const):
(WebCore::FrameTree::findByUniqueName const):
(WebCore::FrameTree::uniqueName const):
(WebCore::FrameTree::uniqueChildName const): Deleted.
(WebCore::FrameTree::generateUniqueName const): Deleted.
* Source/WebCore/page/FrameTree.h:
(WebCore::FrameTree::uniqueName const): Deleted.
(WebCore::FrameTree::resetFrameIdentifiers): Deleted.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::setDocument):

Canonical link: <a href="https://commits.webkit.org/277393@main">https://commits.webkit.org/277393@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c54f053480f5177ef2c99e2024c4d60c7d5e342

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50072 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43437 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24029 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38582 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40838 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19902 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21545 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42005 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5432 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43733 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42425 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51948 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45884 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23694 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44917 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24484 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6688 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23413 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/FrameTree.cpp:
```diff
@@ -537,6 +520,22 @@ unsigned FrameTree::depth() const
+    for (RefPtr frame = top().tree().firstChild(); frame; frame = frame->tree().traverseNext()) {
+        auto frameName = frame->tree().specifiedName();
+        if (frame->tree().specifiedName().isEmpty() || isBlankTargetFrameName(frameName) || frame->tree().childBySpecifiedName(frameName))
+            frameName = makeAtomString("<!--frame"_s, ++frameIndex, "-->"_s);
```
This makes a lot of unnecessary AtomStrings.  Couldn't we just increment frameIndex here and only create one AtomString as we return?

charliewolfe replied:
Yes. Fixed.

achristensen07 commented on Source/WebCore/page/FrameTree.cpp:
```diff
@@ -537,6 +520,22 @@ unsigned FrameTree::depth() const
+    auto frameIndex { 0u };
+    for (RefPtr frame = top().tree().firstChild(); frame; frame = frame->tree().traverseNext()) {
+        auto frameName = frame->tree().specifiedName();
+        if (frame->tree().specifiedName().isEmpty() || isBlankTargetFrameName(frameName) || frame->tree().childBySpecifiedName(frameName))
```
This is a duplicate call to frame->tree().specifiedName()

charliewolfe committed:
Change `FrameTree::uniqueName()` to generate a unique name when called
https://bugs.webkit.org/show_bug.cgi?id=272062
rdar://125808920

Reviewed by Alex Christensen.

Unique frame names currently use a counter to increment a unique frame index each time
`FrameTree::setSpecifiedName` is called. This will not work with site isolation. Instead, we should
change `FrameTree::uniqueName()` to create the unique frame name based on the frame’s position in the
frame tree to mimic previous behavior . Other than layout test output, there are two users of unique
frame names. `HistoryItem` relies on them for session state and to give a consistent identifier for
frame positions after a navigation. Web archives output them in their saved state. Both should continue
to work after this change.

Some tests that output unique frame names needed to be rebased to accommodate this change.

* LayoutTests/fast/forms/form-and-frame-interaction-retains-values-expected.txt:
* LayoutTests/fast/frames/iframe-reparenting-unique-name-expected.txt:
* LayoutTests/fast/frames/reparent-in-unload-contentdocument-expected.txt:
* LayoutTests/http/tests/security/dataURL/xss-DENIED-from-data-url-in-foreign-domain-subframe-expected.txt:
* LayoutTests/http/tests/site-isolation/notify-done-expected.txt:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/page/FrameTree.cpp:
(WebCore::FrameTree::setSpecifiedName):
(WebCore::FrameTree::clearName):
(WebCore::FrameTree::childBySpecifiedName const):
(WebCore::FrameTree::find const):
(WebCore::FrameTree::findByUniqueName const):
(WebCore::FrameTree::uniqueName const):
(WebCore::FrameTree::uniqueChildName const): Deleted.
(WebCore::FrameTree::generateUniqueName const): Deleted.
* Source/WebCore/page/FrameTree.h:
(WebCore::FrameTree::uniqueName const): Deleted.
(WebCore::FrameTree::resetFrameIdentifiers): Deleted.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::setDocument):

Canonical link: https://commits.webkit.org/277393@main
